tools/mpcdump.py

```diff
@@ -0,0 +1,311 @@
+                qst = self._unpack_qstr(ip + 1)
+                print('   ', '0x%02x,' % self.bytecode[ip], qstr_id(qst), '& 0xff,', qstr_id(qst), '>> 8,')
+            else:
+                print('   ', ''.join('0x%02x, ' % self.bytecode[ip + i] for i in range(sz)))
+            ip += sz
+        print('};')
+        print('STATIC const mp_uint_t const_table_data_%s[%u] = {' % (esc_name, len(self.qstrs) + len(self.objs) + len(self.raw_codes)))
+        for qst in self.qstrs:
+            print('    (mp_uint_t)MP_OBJ_NEW_QSTR(%s),' % qstr_id(global_qstrs[qst]))
+        for obj in self.objs:
+            assert 0
+        for n in child_names:
+            print('    (mp_uint_t)&raw_code_%s,' % n)
+        print('};')
+        if simple_name != '<module>':
+            print('STATIC ', end='')
+        print('const mp_raw_code_t raw_code_%s = {' % esc_name)
+        print('    .kind = MP_CODE_BYTECODE,')
+        print('    .scope_flags = 0x%02x,' % prelude[2])
+        print('    .n_pos_args = %u,' % prelude[3])
+        print('    .data.u_byte = {')
+        print('        .bytecode = bytecode_data_%s,' % esc_name)
+        print('        .const_table = const_table_data_%s,' % esc_name)
+        print('        #if MICROPY_PORTABLE_CODE_SAVE')
+        print('        .bc_len = %u,' % len(self.bytecode))
+        print('        .n_obj = %u,' % len(self.objs))
+        print('        .n_raw_code = %u,' % len(self.raw_codes))
+        print('        #endif')
+        print('    },')
+        print('};')
+
+        return esc_name
+
+def read_uint(f):
+    i = 0
+    while True:
+        b = f.read(1)[0]
+        i = (i << 7) | (b & 0x7f)
+        if b & 0x80 == 0:
+            break
+    return i
+
+global_qstrs = []
+def read_qstr(f):
+    ln = read_uint(f)
+    data = str(f.read(ln), 'utf8')
+    global_qstrs.append(data)
+    return len(global_qstrs) - 1
+
+def read_obj(f):
+    assert 0
+
+def read_qstr_and_pack(f, bytecode, ip):
+    qst = read_qstr(f)
+    bytecode[ip] = qst & 0xff
+    bytecode[ip + 1] = qst >> 8
+
+def read_bytecode_qstrs(file, bytecode, ip):
+    while ip < len(bytecode):
+        f, sz = get_opcode_info(bytecode, ip)
+        if f == 1:
+            read_qstr_and_pack(file, bytecode, ip + 1)
+        ip += sz
+
+def read_raw_code(f):
+    bc_len = read_uint(f)
+    bytecode = bytearray(f.read(bc_len))
+    ip, ip2, prelude = extract_prelude(bytecode)
+    read_qstr_and_pack(f, bytecode, ip2) # simple_name
+    read_qstr_and_pack(f, bytecode, ip2 + 2) # source_file
+    read_bytecode_qstrs(f, bytecode, ip)
+    n_obj = read_uint(f)
+    n_raw_code = read_uint(f)
+    qstrs = [read_qstr(f) for _ in range(prelude[3] + prelude[4])]
+    objs = [read_obj(f) for _ in range(n_obj)]
+    raw_codes = [read_raw_code(f) for _ in range(n_raw_code)]
+    return RawCode(bytecode, qstrs, objs, raw_codes)
+
+def read_mpc(filename):
+    with open(filename, 'rb') as f:
+        header = f.read(6)
+        if header[:3] != b'MPC':
```

Comment:
Well, my thinking is along the usual lines of "there can be hundreds of modules, and 3x100 = 300 bytes which can be used for something else". Some basic signature is still nice to have, and one nibble is just enough for hex editor. Or if you really want to support search too, then 1 byte.